src/core/varinfo.jl

```diff
@@ -310,8 +318,23 @@ end
 # Rand & replaying method for VarInfo #
 #######################################
 
-# Check if a vn is set to NULL
-isnan(vi::VarInfo, vn::VarName) = any(isnan.(getval(vi, vn)))
+check_flag(vi::VarInfo, vn::VarName, flag::String) = vi.flags[flag][getidx(vi, vn)]
+set_flag!(vi::VarInfo, vn::VarName, flag::String) = vi.flags[flag][getidx(vi, vn)] = true
+unset_flag!(vi::VarInfo, vn::VarName, flag::String) = vi.flags[flag][getidx(vi, vn)] = false
+
+set_retained_vns_del_by_spl!(vi::VarInfo, spl::Sampler) = begin
+  gidcs = getidcs(vi, spl)
+  if vi.num_produce == 0
+    for i = length(gidcs):-1:1
+      vi.flags["del"][gidcs[i]] = true
+    end
+  else
+    retained = [idx for idx in 1:length(vi.orders) if idx in gidcs && vi.orders[idx] > vi.num_produce]
+    for i = retained
```

Comment:
`i = 1:retained` ?

Reply:
`retained` here is a vector of indices, and `i = retained` is equivalent to `i in retained`. But probably `i in retained` is more readable.